File: pbhhg_py/builtins/functional.py
from typing import Generator, Sequence

from pbhhg_py import abstract_syntax as AS
from pbhhg_py import utils


class Pipe(AS.Function):
    def __init__(self, evaluations: Sequence[AS.Evaluation]):
        super().__init__("Piped ")
        self._evaluations = evaluations

    def __call__(
        self, metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        for evaluation in self._evaluations:
            arg = yield from evaluation(metadata, argv)
            argv = [arg]
        return argv[0]


class Collect(AS.Function):
    def __init__(self, evaluation: AS.Evaluation):
        super().__init__("Collectedly-Receiving ")
        self._evaluation = evaluation

    def __call__(
        self, metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        [seq] = yield from utils.match_arguments(
            metadata, argv, AS.List | AS.ErrorValue, 1
        )
        return (yield from self._evaluation(metadata, seq.value))


class Spread(AS.Function):
    def __init__(self, evaluation: AS.Evaluation):
        super().__init__("Spreadly-Receiving ")
        self._evaluation = evaluation

    def __call__(
        self, metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        argv = [AS.List(tuple(argv))]
        return (yield from self._evaluation(metadata, argv))


def build_tbl(
    proc_functional: utils.ProcFunctional,
) -> dict[str, AS.Evaluation]:
    def _proc(metadata: AS.Metadata):
        def _proc(
            fun: AS.Value,
        ) -> Generator[AS.Value, AS.StrictValue, AS.Evaluation]:
            _fun = yield from utils.strict_functional(metadata, fun)
            return proc_functional(metadata, _fun, general_callable=True)

        return _proc

    def _pipe(
        metadata: AS.Metadata, funs: Sequence[AS.Value]
    ) -> AS.EvalContext:
        evaluations = yield from utils.map_strict_with_hook(
            funs, _proc(metadata)
        )
        return Pipe(evaluations)

    def _collect(
        metadata: AS.Metadata, funs: Sequence[AS.Value]
    ) -> AS.EvalContext:
        utils.check_arity(metadata, funs, 1)
        return Collect((yield from _proc(metadata)(funs[0])))

    def _spread(
        metadata: AS.Metadata, funs: Sequence[AS.Value]
    ) -> AS.EvalContext:
        utils.check_arity(metadata, funs, 1)
        return Spread((yield from _proc(metadata)(funs[0])))

    return {
        "ㄴㄱ": _pipe,  # 연결
        "ㅁㅂ": _collect,  # 모아받기
        "ㅂㅂ": _spread,  # 펴받기
    }

File: pbhhg_js/src/builtins/functional.ts
import * as AS from '../abstractSyntax'
import { checkArity, checkType } from '../utils'

class PipeV extends AS.FunctionV {
  constructor(private funs: AS.Evaluation[]) {
    super('연결된 ')
  }

  execute(
    context: AS.EvalContextBase,
    metadata: AS.Metadata,
    argv: AS.Value[]
  ) {
    const result = this.funs.reduce(
      (argv, f) => [f(context, metadata, argv)],
      argv
    )
    return result[0]
  }
}
class CollectV extends AS.FunctionV {
  constructor(private context: AS.EvalContextBase, private fn: AS.Evaluation) {
    super('모아 받는 ')
  }

  execute(
    context: AS.EvalContextBase,
    metadata: AS.Metadata,
    argv: AS.Value[]
  ) {
    checkArity(metadata, argv, 1)
    const seq = this.context.strict(argv[0])
    const [_seq] = checkType(metadata, [seq], [AS.ListV, AS.ErrorV])
    return this.fn(context, metadata, _seq.value)
  }
}
class SpreadV extends AS.FunctionV {
  constructor(private fn: AS.Evaluation) {
    super('펼쳐 받는 ')
  }

  execute(
    context: AS.EvalContextBase,
    metadata: AS.Metadata,
    argv: AS.Value[]
  ) {
    return this.fn(context, metadata, [new AS.ListV(argv)])
  }
}

function _proc(context: AS.EvalContextBase, metadata: AS.Metadata) {
  return (fun: AS.Value) => context.procFunctional(metadata, fun, true)
}
function _pipe(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  return new PipeV(argv.map(_proc(context, metadata)))
}
function _collect(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkArity(metadata, argv, 1)
  return new CollectV(context, _proc(context, metadata)(argv[0]))
}
function _spread(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkArity(metadata, argv, 1)
  return new SpreadV(_proc(context, metadata)(argv[0]))
}
export default {
  ㄴㄱ: _pipe,
  ㅁㅂ: _collect,
  ㅂㅂ: _spread,
}
